common/types/head_listener.go

```diff
@@ -0,0 +1,22 @@
+package types
+
+import "context"
+
+// HeadListener is a chain agnostic interface that manages connection of Client that receives heads from the blockchain node
+type HeadListener[H Head[BLOCK_HASH], BLOCK_HASH Hashable] interface {
+	// ListenForNewHeads kicks off the listen loop (not thread safe)
+	// done() must be executed upon leaving ListenForNewHeads()
+	ListenForNewHeads(handleNewHead NewHeadHandler[H, BLOCK_HASH], done func())
+
+	// ReceivingHeads returns true if the listener is receiving heads (thread safe)
+	ReceivingHeads() bool
+
+	// Connected returns true if the listener is connected (thread safe)
+	Connected() bool
+
+	// HealthReport returns report of errors within HeadListener
+	HealthReport() map[string]error
+}
+
+// NewHeadHandler is a callback that handles incoming heads
+type NewHeadHandler[H Head[BLOCK_HASH], BLOCK_HASH Hashable] func(ctx context.Context, header H) error
```

Comment:
Will this Head interface work here?
I see currently HeadTracker uses handleNewHead() which actually uses all methods on the commontypes.Head interface, not this one.

See here: https://github.com/smartcontractkit/chainlink/blob/0f05c0ae0c721d4ab8b79fb572c42d0c1f7577d2/core/chains/evm/headtracker/head_tracker.go#L182

Reply:
That makes sense, will be shifting this to common folder instead. Thanks for pointing that out.

Reply:
Good point, thanks for pointing it out. In the future, the HeadTracker would be using the HeadTrackerHead along with other headtracker services. Would generalise that in the next PR on HeadTracker!